### PR TITLE
Use np.asarray to convert a 1-D array to datetime type in array_to_datetime

### PR DESCRIPTION
**Pull request by seisman**
**Description of proposed changes**

Address https://github.com/GenericMappingTools/pygmt/issues/2480.

The initial purpose of this PR is to fix the failing doctests as mentioned in #2480, but in comment https://github.com/GenericMappingTools/pygmt/pull/2481#issuecomment-1497355885 I decided to refactor the `array_to_datetime` function.

---

This PR fixes the following failing test:
```
______________ [doctest] pygmt.clib.conversion.array_to_datetime _______________
276 
277     Examples
278     --------
279     >>> import datetime
280     >>> # numpy.datetime64 array
281     >>> x = np.array(
282     ...     ["2010-06-01", "2011-06-01T12", "2012-01-01T12:34:56"],
283     ...     dtype="datetime64",
284     ... )
285     >>> array_to_datetime(x)
Differences (unified diff with -expected +actual):
    @@ -1,3 +1,3 @@
     DatetimeIndex(['2010-06-01 00:00:00', '2011-06-01 12:00:00',
                    '2012-01-01 12:34:56'],
    -              dtype='datetime64[ns]', freq=None)
    +              dtype='datetime64[s]', freq=None)
```
It's caused by incompatible changes as documented in 
https://pandas.pydata.org/docs/whatsnew/v2.0.0.html#construction-with-datetime64-or-timedelta64-dtype-with-unsupported-resolution.

Fixes #

**Reminders**

- [ ] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If wrapping a new module, open a 'Wrap new GMT module' issue and submit reasonably-sized PRs.
- [ ] If adding new functionality, add an example to docstrings or tutorials.
- [ ] Use underscores (not hyphens) in names of Python files and directories.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

**Comment by seisman** [comment]
A new failing test after the above fix:
```
______________ [doctest] pygmt.clib.conversion.array_to_datetime _______________
303     >>> x = [
304     ...     "2018",
305     ...     "2018-02",
306     ...     "2018-03-01",
307     ...     "2018-04-01T01:02:03",
308     ...     "5/1/2018",
309     ...     "Jun 05, 2018",
310     ...     "2018/07/02",
311     ... ]
312     >>> array_to_datetime(x)
UNEXPECTED EXCEPTION: ValueError('unconverted data remains when parsing with format "%Y": "-02", at position 1. You might want to try:\n    - passing `format` if your strings have a consistent format;\n    - passing `format=\'ISO8601\'` if your strings are all ISO8601 but not necessarily in exactly the same format;\n    - passing `format=\'mixed\'`, and the format will be inferred for each element individually. You might want to use `dayfirst` alongside this.')
Traceback (most recent call last):
  File "/usr/share/miniconda3/envs/pygmt/lib/python3.8/doctest.py", line 1336, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest pygmt.clib.conversion.array_to_datetime[8]>", line 1, in <module>
  File "/home/runner/work/pygmt/pygmt/pygmt/clib/conversion.py", line 329, in array_to_datetime
    return pd.to_datetime(array)
  File "/usr/share/miniconda3/envs/pygmt/lib/python3.8/site-packages/pandas/core/tools/datetimes.py", line 1082, in to_datetime
    result = convert_listlike(argc, format)
  File "/usr/share/miniconda3/envs/pygmt/lib/python3.8/site-packages/pandas/core/tools/datetimes.py", line 453, in _convert_listlike_datetimes
    return _array_strptime_with_fallback(arg, name, utc, format, exact, errors)
  File "/usr/share/miniconda3/envs/pygmt/lib/python3.8/site-packages/pandas/core/tools/datetimes.py", line 484, in _array_strptime_with_fallback
    result, timezones = array_strptime(arg, fmt, exact=exact, errors=errors, utc=utc)
  File "pandas/_libs/tslibs/strptime.pyx", line 530, in pandas._libs.tslibs.strptime.array_strptime
  File "pandas/_libs/tslibs/strptime.pyx", line 355, in pandas._libs.tslibs.strptime.array_strptime
ValueError: unconverted data remains when parsing with format "%Y": "-02", at position 1. You might want to try:
    - passing `format` if your strings have a consistent format;
    - passing `format='ISO8601'` if your strings are all ISO8601 but not necessarily in exactly the same format;
    - passing `format='mixed'`, and the format will be inferred for each element individually. You might want to use `dayfirst` alongside this.
```
caused by the incompatible changes in https://pandas.pydata.org/docs/whatsnew/v2.0.0.html#datetimes-are-now-parsed-with-a-consistent-format.

**Comment by seisman** [comment]
> A new failing test after the above fix:
> 
> ```
> ______________ [doctest] pygmt.clib.conversion.array_to_datetime _______________
> 303     >>> x = [
> 304     ...     "2018",
> 305     ...     "2018-02",
> 306     ...     "2018-03-01",
> 307     ...     "2018-04-01T01:02:03",
> 308     ...     "5/1/2018",
> 309     ...     "Jun 05, 2018",
> 310     ...     "2018/07/02",
> 311     ... ]
> 312     >>> array_to_datetime(x)
> UNEXPECTED EXCEPTION: ValueError('unconverted data remains when parsing with format "%Y": "-02", at position 1. You might want to try:\n    - passing `format` if your strings have a consistent format;\n    - passing `format=\'ISO8601\'` if your strings are all ISO8601 but not necessarily in exactly the same format;\n    - passing `format=\'mixed\'`, and the format will be inferred for each element individually. You might want to use `dayfirst` alongside this.')
> Traceback (most recent call last):
>   File "/usr/share/miniconda3/envs/pygmt/lib/python3.8/doctest.py", line 1336, in __run
>     exec(compile(example.source, filename, "single",
>   File "<doctest pygmt.clib.conversion.array_to_datetime[8]>", line 1, in <module>
>   File "/home/runner/work/pygmt/pygmt/pygmt/clib/conversion.py", line 329, in array_to_datetime
>     return pd.to_datetime(array)
>   File "/usr/share/miniconda3/envs/pygmt/lib/python3.8/site-packages/pandas/core/tools/datetimes.py", line 1082, in to_datetime
>     result = convert_listlike(argc, format)
>   File "/usr/share/miniconda3/envs/pygmt/lib/python3.8/site-packages/pandas/core/tools/datetimes.py", line 453, in _convert_listlike_datetimes
>     return _array_strptime_with_fallback(arg, name, utc, format, exact, errors)
>   File "/usr/share/miniconda3/envs/pygmt/lib/python3.8/site-packages/pandas/core/tools/datetimes.py", line 484, in _array_strptime_with_fallback
>     result, timezones = array_strptime(arg, fmt, exact=exact, errors=errors, utc=utc)
>   File "pandas/_libs/tslibs/strptime.pyx", line 530, in pandas._libs.tslibs.strptime.array_strptime
>   File "pandas/_libs/tslibs/strptime.pyx", line 355, in pandas._libs.tslibs.strptime.array_strptime
> ValueError: unconverted data remains when parsing with format "%Y": "-02", at position 1. You might want to try:
>     - passing `format` if your strings have a consistent format;
>     - passing `format='ISO8601'` if your strings are all ISO8601 but not necessarily in exactly the same format;
>     - passing `format='mixed'`, and the format will be inferred for each element individually. You might want to use `dayfirst` alongside this.
> ```
> 
> It's caused by the incompatible changes in https://pandas.pydata.org/docs/whatsnew/v2.0.0.html#datetimes-are-now-parsed-with-a-consistent-format.
> 
> > In the past, [to_datetime()](https://pandas.pydata.org/docs/reference/api/pandas.to_datetime.html#pandas.to_datetime) guessed the format for each element independently. This was appropriate for some cases where elements had mixed date formats - however, it would regularly cause problems when users expected a consistent format but the function would switch formats between elements. As of version 2.0.0, parsing will use a consistent format, determined by the first non-NA value (unless the user specifies a format, in which case that is used).
> 
> In our case, we're passing datetimes in different formats, but pandas tries to parse all datetimes using a consistent format `%Y`, which is guessed based on the first datetime (i.e., `2018`).

Here are some possible options:

1. Always passing `format="mixed"` to the `pd.to_datetime` callhttps://github.com/GenericMappingTools/pygmt/blob/aaa0ca006d3d1fddb86db3ed8a9250fc0e1b9597/pygmt/clib/conversion.py#L329 so that the format will be inferred for each element individually. This is the default behavior in pandas < 2.0.0. However, `format="mixed"` is not available for pandas < 2.0.0, so we have to pin pandas to >=2.0.0 or check the pandas version.
2. Just remove the failing test, which also means we no longer support a list of datetimes with mixed formats.
3. Let `array_to_datetime` support `**kwargs` so that we can pass any keyword arguments to `pd.to_dataframe`. But the `array_to_datetime` function is only internally used by PyGMT, so users who pass a list of datetimes with mixed formats will still see the same error.

I'm inclined to option 2. What do you think?

**Comment by seisman** [comment]
The `array_to_datetime` function was added in PR https://github.com/GenericMappingTools/pygmt/pull/464 to support various datetime types and is used only once in the `put_vector` function:
https://github.com/GenericMappingTools/pygmt/blob/aaa0ca006d3d1fddb86db3ed8a9250fc0e1b9597/pygmt/clib/session.py#L797-L805

Please note that before calling `array_to_datetime(vector)`, we also call `_check_dtype_and_dim(vector)` (at line 797). 

In the `_check_dtype_and_dim` function, we use `np.asarray(array, dtype=np.datetime64)` to check if a list/array can be recognized as `np.datetime64` (see below).

https://github.com/GenericMappingTools/pygmt/blob/aaa0ca006d3d1fddb86db3ed8a9250fc0e1b9597/pygmt/clib/session.py#L744-L752

Thus, if a list/vector can't be processed by `np.asarray(array, dtype=np.datetime64)`, it will not be recognized as a valid datetime list/vector and won't be processed by `array_to_datearray`.

In short, I think we should use `np.asarray(array, dtype=np.datetime64)` instead of `pd.to_datetime()` to do the conversion.

File: pygmt/clib/conversion.py [hunk 1]
```diff
@@ -272,58 +270,57 @@ def array_to_datetime(array):
...
 
     >>> # Raw datetime strings in various format
     >>> x = [
     ...     "2018",
     ...     "2018-02",
     ...     "2018-03-01",
     ...     "2018-04-01T01:02:03",
-    ...     "5/1/2018",
-    ...     "Jun 05, 2018",
-    ...     "2018/07/02",
```
Comment (seisman) [inline pygmt/clib/conversion.py:310]: These three types of datetime inputs are no longer supported. Actually, they're never supported because they won't be recognized as valid datetime strings in the `_check_dtype_and_dim` function.

**Comment by seisman** [comment]
> In short, I think we should use `np.asarray(array, dtype=np.datetime64)` instead of `pd.to_datetime()` to do the conversion.

I've made the change in commit 608e29bf59ea1340b12971c39fe6083dc11f15c2 and also use the same function in `_check_dtype_and_dim` (commit 2ac723efaa81bfc170386aac5b665cb2474ac8e4). Now the tests pass.

File: pygmt/clib/conversion.py [hunk 2]
```diff
@@ -252,10 +251,9 @@ def kwargs_to_ctypes_array(argument, kwargs, dtype):
 
 def array_to_datetime(array):
     """
-    Convert an 1-D datetime array from various types into pandas.DatetimeIndex
-    (i.e., numpy.datetime64).
+    Convert an 1-D datetime array from various types into numpy.datetime64.
```
Comment (weiji14) [inline pygmt/clib/conversion.py:254]:
```suggestion
    Convert a 1-D datetime array from various types into numpy.datetime64.
```

File: pygmt/clib/conversion.py [hunk 3]
```diff
@@ -272,58 +270,58 @@ def array_to_datetime(array):
...
     ... ]
-    >>> array_to_datetime(x)  # doctest: +NORMALIZE_WHITESPACE
-    DatetimeIndex(['2018-01-01', '2018-01-01', '2018-01-01'],
-                  dtype='datetime64[ns]', freq=None)
+    >>> array_to_datetime(x)
+    array(['2018-01-01T00:00:00.000000', '2018-01-01T00:00:00.000000',
+           '2018-01-01T00:00:00.000000'], dtype='datetime64[us]')
     """
-    return pd.to_datetime(array)
+    return np.asarray(array, dtype=np.datetime64)
```
Comment (weiji14) [inline pygmt/clib/conversion.py:327]: I feel like there was a reason we used `pd.to_datetime` instead of `np.asarray` in https://github.com/GenericMappingTools/pygmt/pull/464, maybe to support funny dates like 'Jun 05, 2018', but since the tests pass and you mentioned in https://github.com/GenericMappingTools/pygmt/pull/2481/files#r1158514955 that it's not actually supported by `_check_dtype_and_dim`, then it should be fine :crossed_fingers:

**Review by weiji14 — APPROVED** [review]

File: pygmt/clib/conversion.py (hunk 3 above)
Comment (seisman) [reply pygmt/clib/conversion.py:327]: > I feel like there was a reason we used `pd.to_datetime` instead of `np.asarray` in #464, maybe to support funny dates like 'Jun 05, 2018',

In the [Plotting datetime charts tutorial](https://pygmt.org/dev/tutorials/advanced/date_time_charts.html), we probably need to add a paragraph or a subsection to explain that, for non-ISO datetimes like 'Jun 05, 2018', people should use `pd.to_datetime` to process it first before passing it to PyGMT.

Comment (weiji14) [reply pygmt/clib/conversion.py:327]: Under https://www.pygmt.org/v0.9.0/tutorials/advanced/date_time_charts.html#generating-an-automatic-region, `pd.to_datetime` is used to convert dates like "20220712" into a recognizable format. Maybe add a few sentences somewhere there?

Comment (seisman) [reply pygmt/clib/conversion.py:327]: I added a sentence under the "Using ISO Format" section (ada56160140b4ea03c50f2c6436450b98c1f4bca).

Preview: https://pygmt-dev--2481.org.readthedocs.build/en/2481/tutorials/advanced/date_time_charts.html#using-iso-format

**Comment by seisman** [comment]
For mysterious reasons, suddenly `mamba` now installs xarray v0.19.0 which was released more than 1 year ago, and doesn't work with pandas 2.0.0.

**Comment by seisman** [comment]
> For mysterious reasons, suddenly `mamba` now installs xarray v0.19.0 which was released more than 1 year ago, and doesn't work with pandas 2.0.0.

Related to https://github.com/pydata/xarray/issues/7716.

**Comment by weiji14** [comment]
> > For mysterious reasons, suddenly `mamba` now installs xarray v0.19.0 which was released more than 1 year ago, and doesn't work with pandas 2.0.0.
> 
> Related to [pydata/xarray#7716](https://github.com/pydata/xarray/issues/7716).

Yeah, looks like they did a repodata patch at https://github.com/conda-forge/conda-forge-repodata-patches-feedstock/pull/426 for xarray versions 2022.10.0 to 2023.2.0 at (i.e. the CalVer versions), but they missed the older SemVer versions (e.g. 0.19.0) so we'll need to wait for https://github.com/conda-forge/conda-forge-repodata-patches-feedstock/pull/429 :shrug:

File: examples/tutorials/advanced/date_time_charts.py [hunk 4]
```diff
@@ -124,6 +124,13 @@
 )
 fig.show()
 
+###############################################################################
+#
+# PyGMT doesn't recognize non-ISO datetime strings like "Jun 05, 2018". If your
+# data contain non-ISO datetime strings, you can convert them to a recognized
+# format using :func:`pandas.to_datetime` and then pass to PyGMT.
```
Comment (michaelgrund) [inline examples/tutorials/advanced/date_time_charts.py:131]:
```suggestion
# format using :func:`pandas.to_datetime` and then pass it to PyGMT.
```

File: pygmt/clib/conversion.py [hunk 5]
```diff
@@ -252,10 +251,9 @@ def kwargs_to_ctypes_array(argument, kwargs, dtype):
 
 def array_to_datetime(array):
     """
-    Convert an 1-D datetime array from various types into pandas.DatetimeIndex
-    (i.e., numpy.datetime64).
+    Convert n 1-D datetime array from various types into numpy.datetime64.
```
Comment (michaelgrund) [inline pygmt/clib/conversion.py:254]:
```suggestion
    Convert an 1-D datetime array from various types into numpy.datetime64.
```
